File: joao/0552_student_attendance_record_ii.go
const MOD = 1_000_000_007

func checkRecord(n int) int {
	dp := initializeDP(n)
	populateDP(dp, n)
	return sumValidStates(dp, n)
}

func initializeDP(n int) [][][]int {
	dp := make([][][]int, n+1)
	for i := range dp {
		dp[i] = make([][]int, 2)
		for j := range dp[i] {
			dp[i][j] = make([]int, 3)
		}
	}
	dp[0][0][0] = 1
	return dp
}

func populateDP(dp [][][]int, n int) {
	for i := 1; i <= n; i++ {
		for j := 0; j < 2; j++ {
			for k := 0; k < 3; k++ {
				dp[i][j][0] = (dp[i][j][0] + dp[i-1][j][k]) % MOD

				if j < 1 {
					dp[i][j+1][0] = (dp[i][j+1][0] + dp[i-1][j][k]) % MOD
				}

				if k < 2 {
					dp[i][j][k+1] = (dp[i][j][k+1] + dp[i-1][j][k]) % MOD
				}
			}
		}
	}
}

func sumValidStates(dp [][][]int, n int) int {
	result := 0
	for j := 0; j < 2; j++ {
		for k := 0; k < 3; k++ {
			result = (result + dp[n][j][k]) % MOD
		}
	}
	return result
}

File: joao/python/0552_student_attendance_record_ii.py
class Solution:
    def checkRecord(self, n: int) -> int:
        MOD = 10**9 + 7

        memo = {}

        def count_records(n, A_count, L_count):
            if n == 0:
                return 1

            if (n, A_count, L_count) in memo:
                return memo[(n, A_count, L_count)]

            result = 0

            result += count_records(n - 1, A_count, 0)
            result %= MOD

            if A_count < 1:
                result += count_records(n - 1, A_count + 1, 0)
                result %= MOD

            if L_count < 2:
                result += count_records(n - 1, A_count, L_count + 1)
                result %= MOD

            memo[(n, A_count, L_count)] = result
            return result

        return count_records(n, 0, 0)
